File: crates/bv/src/fixup.rs
use std::collections::HashSet;

use crate::syntax::{BasicNode, Function, FunctionBody, Node, NodeAddr, NodeId};

impl Function {
    pub(crate) fn fixup(&mut self) {
        if let Some(body) = self.body_mut() {
            body.ensure_simple_entrypoint();
            body.ensure_closed();
        }
    }
}

impl FunctionBody {
    fn fixup(&mut self) {
        self.ensure_simple_entrypoint();
        self.ensure_closed();
    }

    fn ensure_simple_entrypoint(&mut self) {
        let addr = self.fresh_node_addr();
        self.nodes.insert(
            addr,
            Node::Basic(BasicNode {
                next: self.entry_point.clone(),
                var_updates: vec![],
            }),
        );
        self.entry_point = NodeId::Addr(addr);
    }

    fn fresh_node_addr(&self) -> NodeAddr {
        // implementation matches graph_refine.syntax.fresh_node
        let hint = 1;
        let mut n = hint;
        n = (n | 15) + 2;
        while self.nodes.contains_key(&n) {
            n += 16;
        }
        n
    }

    fn ensure_closed(&mut self) {
        let mut dead = HashSet::new();
        for node in self.nodes.values() {
            for cont in node.conts() {
                if let NodeId::Addr(cont_addr) = cont {
                    if !self.nodes.contains_key(&cont_addr) {
                        dead.insert(cont_addr);
                    }
                }
            }
        }
        for addr in dead {
            self.nodes.insert(
                addr,
                Node::Basic(BasicNode {
                    next: NodeId::Err,
                    var_updates: vec![],
                }),
            );
        }
    }
}

File: hs/components/core/BV/Core/Stages/Fixup.hs
module BV.Core.Stages.Fixup
    ( fixupProgram
    ) where

import Control.Monad (forM_)
import Control.Monad.State
import Data.Foldable (find)
import qualified Data.Map as M
import Data.Maybe (fromJust)
import qualified Data.Set as S
import Optics

import BV.Core.Types
import BV.Core.Types.Extras

fixupProgram :: Program -> Program
fixupProgram = #functions % traversed % #body % traversed %~ fixupFunctionBody

fixupFunctionBody :: FunctionBody -> FunctionBody
fixupFunctionBody = execState $ do
    ensureSimpleEntrypoint
    ensureClosed

ensureSimpleEntrypoint :: State FunctionBody ()
ensureSimpleEntrypoint = do
    origEntryPointId <- use #entryPoint
    newEntryPointAddr <- use $ #nodes % to freshNodeAddr
    modify $ #nodes % at newEntryPointAddr ?~ trivialNode origEntryPointId
    assign #entryPoint (Addr newEntryPointAddr)

ensureClosed :: State FunctionBody ()
ensureClosed = do
    allContAddrs <- gets $ S.fromList . toListOf (#nodes % traversed % nodeConts % #_Addr)
    liveAddrs <- use $ #nodes % to M.keysSet
    let deadAddrs = allContAddrs `S.difference` liveAddrs
    forM_ deadAddrs $ \deadAddr -> do
        modify $ #nodes % at deadAddr ?~ trivialNode Err

-- implementation matches graph_refine.syntax.fresh_node
freshNodeAddr :: NodeMap -> NodeAddr
freshNodeAddr nodeMap = fromJust . find (`M.notMember` nodeMap) $
    iterate (+ 16) 17
